### DISCUSSION TITLE
Allow shader global uniforms to be arrays

### DISCUSSION THREAD

**Discussion by ramblingstranger** [godotengine/godot-proposals#9553] [3D]
Right now, Godot allows regular uniforms to be arrays. See the PR when this was implemented: [Added support for uniform arrays in shaders](https://github.com/godotengine/godot/pull/49485).

However, it still doesn't allow you to do this for global uniforms. The docs list out all of the possible [global shader parameter types](https://docs.godotengine.org/en/stable/classes/class_renderingserver.html#enum-renderingserver-globalshaderparametertype), and arrays don't appear to be included. You can also see this limitation when adding a shader global in the project settings dialog.

![image](https://github.com/godotengine/godot-proposals/assets/114190975/c269f7bf-4de6-41f5-bf53-1f5f06347d62)
> No options for arrays of any kind.

Would it be possible to add arrays as a global uniform type?

**Comment by tripodsan** [comment]
FWIW, I added a "workaround" with a `sampler2D`. for my use case, I want to pass an array of `vec3` of node positions in order to calculate fake shadows. here's what I came up with (I removed some of the sorting and filtering code for clarity):

```swift
extends Node

const MAX_ENTITY_SHADOWS = 512

const GLOBAL_UNIFORM_ENTITY_POSITIONS := &"entity_positions"

var _texture_rid:RID

var _pos_buffer:PackedFloat32Array = []

var _rd:RenderingDevice

func _ready() -> void:
  if not _texture_rid.is_valid():
    _pos_buffer.resize(MAX_ENTITY_SHADOWS * 4)
    _pos_buffer.fill(0.0)
    RenderingServer.call_on_render_thread(_initialize_rendering)

func update_entity_positions(player:Node3D, npes:Array[Node3D])->void:
  if not _texture_rid.is_valid(): return
  # first vec4 contains the player cords, and the number of entities in W
  _pos_buffer[0] = player.global_position.x
  _pos_buffer[1] = player.global_position.y
  _pos_buffer[2] = player.global_position.z
  _pos_buffer[3] = npes.size() + 1 # include player
  var idx:int = 4
  for e:Entity in npes:
    _pos_buffer[idx + 0] = e.global_position.x
    _pos_buffer[idx + 1] = e.global_position.y
    _pos_buffer[idx + 2] = e.global_position.z
    idx += 4
  _rd.texture_update(_texture_rid, 0, _pos_buffer.to_byte_array())

func _initialize_rendering() -> void:
  _rd = RenderingServer.get_rendering_device()

  # Create the texture for the positions
  var tf := RDTextureFormat.new()
  tf.format = RenderingDevice.DATA_FORMAT_R32G32B32A32_SFLOAT
  tf.texture_type = RenderingDevice.TEXTURE_TYPE_2D
  tf.width = MAX_ENTITY_SHADOWS
  tf.height = 1
  tf.depth = 1
  tf.array_layers = 1
  tf.mipmaps = 1
  tf.usage_bits = (
      RenderingDevice.TEXTURE_USAGE_SAMPLING_BIT |
      RenderingDevice.TEXTURE_USAGE_CAN_UPDATE_BIT
    )
  _texture_rid = _rd.texture_create(tf, RDTextureView.new(), [])
  var texture := Texture2DRD.new()
  texture.texture_rd_rid = _texture_rid
  RenderingServer.global_shader_parameter_set(GLOBAL_UNIFORM_ENTITY_POSITIONS, texture)
```

define the global uniform as Texture2DRD:

<img width="658" height="264" alt="image" src="https://github.com/user-attachments/assets/cb4b1127-04f6-4316-8462-6d81033fef81" />

and add the script as autoload:
<img width="609" height="31" alt="image" src="https://github.com/user-attachments/assets/b877149d-4fb6-4a69-80dd-b0cd698b42f9" />

and in the shader use it as follows:

```glsl
global uniform sampler2D entity_positions;

void fragment() {
 ...
  // calculate entity shadow
  int num_entities = int(texelFetch(entity_positions, ivec2(0), 0).w);
  
  for (int idx = 0; idx < num_entities; idx++) {
    vec3 entity_pos = texelFetch(entity_positions, ivec2(idx, 0), 0).xyz;
    ....
  }
  
}
```

### Notes
- call `ShadowController.update_entity_positions()` whenever the positions change.

**Reply by Brandon-M-Hartman** [reply]
> FWIW, I added a "workaround" with a `sampler2D`. for my use case, I want to pass an array of `vec3` of node positions in order to calculate fake shadows. here's what I came up with (I removed some of the sorting and filtering code for clarity):
> 
> ```swift
> extends Node
> 
> const MAX_ENTITY_SHADOWS = 512
> 
> const GLOBAL_UNIFORM_ENTITY_POSITIONS := &"entity_positions"
> 
> var _texture_rid:RID
> 
> var _pos_buffer:PackedFloat32Array = []
> 
> var _rd:RenderingDevice
> 
> func _ready() -> void:
>   if not _texture_rid.is_valid():
>     _pos_buffer.resize(MAX_ENTITY_SHADOWS * 4)
>     _pos_buffer.fill(0.0)
>     RenderingServer.call_on_render_thread(_initialize_rendering)
> 
> func update_entity_positions(player:Node3D, npes:Array[Node3D])->void:
>   if not _texture_rid.is_valid(): return
>   # first vec4 contains the player cords, and the number of entities in W
>   _pos_buffer[0] = player.global_position.x
>   _pos_buffer[1] = player.global_position.y
>   _pos_buffer[2] = player.global_position.z
>   _pos_buffer[3] = npes.size() + 1 # include player
>   var idx:int = 4
>   for e:Entity in npes:
>     _pos_buffer[idx + 0] = e.global_position.x
>     _pos_buffer[idx + 1] = e.global_position.y
>     _pos_buffer[idx + 2] = e.global_position.z
>     idx += 4
>   _rd.texture_update(_texture_rid, 0, _pos_buffer.to_byte_array())
> 
> func _initialize_rendering() -> void:
>   _rd = RenderingServer.get_rendering_device()
> 
>   # Create the texture for the positions
>   var tf := RDTextureFormat.new()
>   tf.format = RenderingDevice.DATA_FORMAT_R32G32B32A32_SFLOAT
>   tf.texture_type = RenderingDevice.TEXTURE_TYPE_2D
>   tf.width = MAX_ENTITY_SHADOWS
>   tf.height = 1
>   tf.depth = 1
>   tf.array_layers = 1
>   tf.mipmaps = 1
>   tf.usage_bits = (
>       RenderingDevice.TEXTURE_USAGE_SAMPLING_BIT |
>       RenderingDevice.TEXTURE_USAGE_CAN_UPDATE_BIT
>     )
>   _texture_rid = _rd.texture_create(tf, RDTextureView.new(), [])
>   var texture := Texture2DRD.new()
>   texture.texture_rd_rid = _texture_rid
>   RenderingServer.global_shader_parameter_set(GLOBAL_UNIFORM_ENTITY_POSITIONS, texture)
> ```
> 
> define the global uniform as Texture2DRD:
> <img alt="image" width="658" height="264" src="https://private-user-images.githubusercontent.com/917628/540262281-cb4b1127-04f6-4316-8462-6d81033fef81.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMyNDI4NzQsIm5iZiI6MTc3MzI0MjU3NCwicGF0aCI6Ii85MTc2MjgvNTQwMjYyMjgxLWNiNGIxMTI3LTA0ZjYtNDMxNi04NDYyLTZkODEwMzNmZWY4MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMxMVQxNTIyNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMTBjNGNmNDVkNzZhNTkyZjU2ZjgxNzFjYmNlMTU4ZWM3NGUwZjY5NzNjNWU4YjA1ZGIxZDFhMjkxNGQ2ZjRlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.r3eZnNFDlCJHuEkxZiBEVqtowqz4WGY450RMF21Fxgk">
> 
> and add the script as autoload: <img alt="image" width="609" height="31" src="https://private-user-images.githubusercontent.com/917628/540263374-b877149d-4fb6-4a69-80dd-b0cd698b42f9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzMyNDI4NzQsIm5iZiI6MTc3MzI0MjU3NCwicGF0aCI6Ii85MTc2MjgvNTQwMjYzMzc0LWI4NzcxNDlkLTRmYjYtNGE2OS04MGRkLWIwY2Q2OThiNDJmOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMxMVQxNTIyNTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zODI2ZThmMmQxNzAyOTRjYTdhYTk3MmYxYmI0YWZkYzc3OWJmY2JkMmViZTY0NzFiYTIyNzM0ZWIyNmFjZjUyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.l3YqkakWpWFGZ2m525yYxDWQ8-ahM3wuZJ3NHKxK1a4">
> 
> and in the shader use it as follows:
> 
> ```glsl
> global uniform sampler2D entity_positions;
> 
> void fragment() {
>  ...
>   // calculate entity shadow
>   int num_entities = int(texelFetch(entity_positions, ivec2(0), 0).w);
>   
>   for (int idx = 0; idx < num_entities; idx++) {
>     vec3 entity_pos = texelFetch(entity_positions, ivec2(idx, 0), 0).xyz;
>     ....
>   }
>   
> }
> ```
> 
> ### Notes
> 
>     * call `ShadowController.update_entity_positions()` whenever the positions change.

Would you be willing to share the full code? I'm interested in the sorting and filtering capabilities. Trying to solve this same use case myself and appreciate your solution here a lot!

**Reply by tripodsan** [reply]
it's not much, just:

```
const MAX_RENDER_DISTANCE = 50
.
.
  npes = npes.filter(func (npe:Entity): return npe.distance_to_player <= MAX_RENDER_DISTANCE)
  if npes.size() >= MAX_ENTITY_SHADOWS - 1:
    npes.sort_custom(func (e0:Entity, e1:Entity): return e0.distance_to_player < e1.distance_to_player)
    npes.resize(MAX_ENTITY_SHADOWS - 1)
  num_entities = npes.size()

```
